api/controller.py

```diff
@@ -514,6 +514,37 @@ def annotator(self, lane, facets=None, *args, **kwargs):
         if not library:
             return CirculationManagerAnnotator(lane)
 
+        # When we know that the request was initiated by an authenticated patron
+        # we need to pass dictionaries containing loans, hold,
+        # and fulfilment objects related to the current work.
+        # They will be used by the annotator to generate fulfilment, revoke links, etc.
+        work = kwargs.pop('work', None)
+        patron = kwargs.pop('patron', None)
+
+        if work and patron:
+            active_loans_by_work = {
+                work: loan
+                for license_pool in work.license_pools
+                for loan in license_pool.loans
```

Comment:
This is going to map a Work to a list of all the loans issued for the Work to _any_ patron -- this could be thousands of loans. I wouldn't expect this to work and it's definitely inefficient. `active_loans_by_work` needs to contain a dictionary that maps a Work object to the Loan for the authenticated patron only. And this whole thing should happen only when it's necessary--putting it in `CirculationManager.annotator` will make it happen whenever there is an authenticated patron.

Reply:
I removed all of this logic and used `LibraryLoanAndHoldAnnotator` instead.